CoseHandler/CoseHandler.cs

```diff
@@ -591,27 +592,46 @@ public static class CoseHandler
         if (publicKey is null)
         {
             errorCodes.Add(ValidationFailureCode.NoPublicKey);
-            return new ValidationResult(false, errorCodes, certValidationResults, chain);
+            return new ValidationResult(false, errorCodes, certValidationResults, chain, validationType: ContentValidationType.ContentValidationNotPerformed);
         }
 
         // Validate that the COSE header is formatted correctly and that the payload and hash are consistent.
         bool messageVerified = false;
+
+        // check for external payload
+        bool hasBytes = !payloadBytes.IsNullOrEmpty();
+        bool hasStream = payloadStream is not null;
+
+        // Check for an indirect signature, where the content header contains the hash of the payload, and the algorithm is stored in the message.
+        // If this is the case and external content is provided, we can validate an external payload hash against the hash stored in the cose message content.
+        bool indirectSignature = msg.IsDetachedSignature();
+
+        // Determine the type of content validation to perform.
+        ContentValidationType cvt = (indirectSignature && (hasBytes || hasStream)) ? ContentValidationType.Indirect :
+            hasBytes || hasStream ? ContentValidationType.Detached : ContentValidationType.Embedded;
+
         try
         {
-            if (!payloadBytes.IsNullOrEmpty())
-            {
-                // Detached payload received as byte array
-                messageVerified = msg.VerifyDetached(publicKey, new ReadOnlySpan<byte>(payloadBytes));
-            }
-            else if (payloadStream is not null)
-            {
-                // Detached payload received as a stream
-                messageVerified = Task.Run(() => msg.VerifyDetachedAsync(publicKey, payloadStream)).GetAwaiter().GetResult();
-            }
-            else
+            switch (cvt)
             {
-                // Embedded payload
-                messageVerified = msg.VerifyEmbedded(publicKey);
+                // Indirect signature validation. Validate external payload hash against embedded hash + Embedded signature validation.
+                case ContentValidationType.Indirect:
+                    messageVerified = hasBytes ?
+                        (msg.VerifyEmbedded(publicKey) && msg.SignatureMatches(payloadBytes)) :
+                        (msg.VerifyEmbedded(publicKey) && msg.SignatureMatches(payloadStream));
+                    break;

```

Comment:
I'm thinking about the case where the signature is indirect but the user fails to supply payload. 
In this flow, it will be validated as a normal embedded signature. Have you tested this to see the outcome?
I'm thinking we probably should throw an informative exception here so the user knows what's going on instead of failing on a valid signature. You might need to add an error code for this as well, 
...and update CoseSignTool.ValidateCommand to both return an informative error and mention the requirement for payload with indirect signatures in the command line help.

Reply:
Yeah this was something that I went back and forth on for a while. The thinking here is that we would provide a way for users to do partial validation of indirect signatures when the payload is not available. Perhaps this is not a reasonable scenario.

Reply:
Oh, okay. I think having a partial validation option is reasonable, but we should be secure by default, so turn it on with a flag, and carry the flag into CoseSignTool as well.